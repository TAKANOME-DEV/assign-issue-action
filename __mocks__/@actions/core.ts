// import {InputOptions} from '@actions/core';

export const getInput = jest.fn().mockImplementation((name: string) => name);
// jest.spyOn(core, 'getInput').mockImplementation((name: string) => {
//   return inputs[name]
// })
// export const getInput = jest
//   .spyOn(core, 'getInput')
//   .mockImplementation((name, options) => {
//     return jest.requireActual('@actions/core').getInput(name, options);
//   });
