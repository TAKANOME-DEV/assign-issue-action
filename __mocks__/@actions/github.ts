export const context = {
  payload: {
    comment: {
      user: {
        login: 'John',
      },
      body: '/assign-to-me',
      html_url: 'https://github.com/TAKANOME-DEV/testing',
    },
    issue: {
      labels: [],
      assignee: null,
      number: 1,
    },
  },
  repo: {
    owner: 'TAKANOME-DEV',
    repo: 'testing',
  },
  action: 'created',
};

const mockApi = {
  rest: {
    issues: {
      addAssignees: jest.fn(),
      addLabels: jest.fn(),
      createComment: jest.fn(),
      removeLabel: jest.fn(),
      removeAssignee: jest.fn(),
    },
    search: {
      issuesAndPullRequests: jest.fn().mockImplementation(() => ({
        status: 200,
        body: {
          items: [
            {
              number: 1,
              labels: [],
              assignee: {
                login: 'John',
              },
            },
          ],
        },
      })),
    },
  },
};

export const getOctokit = jest.fn().mockImplementation(() => mockApi);
