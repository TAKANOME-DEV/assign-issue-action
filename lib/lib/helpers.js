"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@actions/core");
const fs_1 = __importDefault(require("fs"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const path_1 = __importDefault(require("path"));
function getInputs() {
    return {
        assigned_label: (0, core_1.getInput)('assigned_label'),
        required_label: (0, core_1.getInput)('required_label'),
        pin_label: (0, core_1.getInput)('pin_label'),
        days_until_unassign: parseInt((0, core_1.getInput)('days_until_unassign'), 10),
        stale_assignment_label: (0, core_1.getInput)('stale_assignment_label'),
        assigned_comment: (0, core_1.getInput)('assigned_comment'),
        github_token: (0, core_1.getInput)('github_token'),
    };
}
/**
 * Helper that reads the `action.yml` and includes the default values
 * for each input as an environment variable, like the Actions runtime does.
 */
function getDefaultValues() {
    const yaml = fs_1.default.readFileSync(
    // eslint-disable-next-line no-undef
    path_1.default.join(__dirname, '../../action.yml'), 'utf8');
    const { inputs } = js_yaml_1.default.load(yaml);
    return Object.keys(inputs).reduce((acc, key) => {
        if ('default' in inputs[key]) {
            return Object.assign(Object.assign({}, acc), { [`INPUT_${key.toUpperCase()}`]: inputs[key].default });
        }
        else {
            return acc;
        }
    }, {});
}
exports.default = {
    getInputs,
    getDefaultValues,
};
