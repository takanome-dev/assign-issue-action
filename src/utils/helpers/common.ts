/**
 * Utility function to get a date string in the format YYYY-MM-DD
 * @param days - The number of days to subtract from the current date
 * @returns A date string in the format YYYY-MM-DD
 */
export function since(days: number) {
  const totalDaysInMilliseconds = days * 24 * 60 * 60 * 1000;
  const date = new Date(+new Date() - totalDaysInMilliseconds);

  return new Date(date).toISOString().substring(0, 10);
}

/**
 * Utility function to split array into chunks
 * @param array - The array to split
 * @param chunkSize - The size of each chunk
 * @returns An array of arrays
 */
export function chunkArray<T>(array: T[], chunkSize: number): T[][] {
  const chunks = [];
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
}

/**
 * Utility function to calculate days between dates
 * @param start - The start date
 * @param end - The end date
 * @returns The number of days between the two dates
 */
export function getDaysBetween(start: Date, end: Date) {
  const diffTime = Math.abs(end.getTime() - start.getTime());
  return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
}
