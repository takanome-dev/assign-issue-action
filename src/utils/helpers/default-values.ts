import fs from 'fs';
import jsYaml from 'js-yaml';
import path from 'path';

/**
 * Helper that reads the `action.yml` and includes the default values
 * for each input as an environment variable, like the Actions runtime does.
 */
export function getDefaultValues() {
  const yaml = fs.readFileSync(
    path.join(__dirname, '../../../action.yml'),
    'utf8',
  );
  const { inputs } = jsYaml.load(yaml) as any;

  return Object.keys(inputs).reduce((acc, key) => {
    if ('default' in inputs[key]) {
      return {
        ...acc,
        [`INPUT_${key.toUpperCase()}`]: inputs[key].default,
      };
    } else {
      return acc;
    }
  }, {});
}
