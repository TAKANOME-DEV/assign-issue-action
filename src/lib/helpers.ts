import { getInput } from '@actions/core';
import fs from 'fs';
import jsYaml from 'js-yaml';
import path from 'path';
import { Inputs } from '../types';

function getInputs(): Inputs {
  return {
    assigned_label: getInput('assigned_label'),
    required_label: getInput('required_label'),
    pin_label: getInput('pin_label'),
    days_until_unassign: parseInt(getInput('days_until_unassign'), 10),
    stale_assignment_label: getInput('stale_assignment_label'),
    assigned_comment: getInput('assigned_comment'),
    github_token: getInput('github_token'),
  };
}

/**
 * Helper that reads the `action.yml` and includes the default values
 * for each input as an environment variable, like the Actions runtime does.
 */
function getDefaultValues() {
  const yaml = fs.readFileSync(
    // eslint-disable-next-line no-undef
    path.join(__dirname, '../../action.yml'),
    'utf8',
  );
  const { inputs } = jsYaml.load(yaml) as any;

  return Object.keys(inputs).reduce((acc, key) => {
    if ('default' in inputs[key]) {
      return {
        ...acc,
        [`INPUT_${key.toUpperCase()}`]: inputs[key].default,
      };
    } else {
      return acc;
    }
  }, {});
}

export default {
  getInputs,
  getDefaultValues,
};
