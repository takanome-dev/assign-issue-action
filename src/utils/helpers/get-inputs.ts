import { getInput } from '@actions/core';
import { INPUTS } from '../lib/inputs';

export function getInputs() {
  return {
    assigned_label: getInput(INPUTS.ASSIGNED_LABEL),
    required_label: getInput(INPUTS.REQUIRED_LABEL),
    pin_label: getInput(INPUTS.PIN_LABEL),

    days_until_unassign: parseInt(getInput(INPUTS.DAYS_UNTIL_UNASSIGN), 10),
    stale_assignment_label: getInput(INPUTS.STALE_ASSIGNMENT_LABEL),
    assigned_comment: getInput(INPUTS.ASSIGNED_COMMENT),
    assignment_suggestion_comment: getInput(
      INPUTS.ASSIGNMENT_SUGGESTION_COMMENT,
    ),
    github_token: getInput(INPUTS.GITHUB_TOKEN),
    maintainers: getInput(INPUTS.MAINTAINERS),
    enable_auto_suggestion: getInput(INPUTS.ENABLE_AUTO_SUGGESTION),

    self_assign_cmd: getInput(INPUTS.SELF_ASSIGN_CMD),
    self_unassign_cmd: getInput(INPUTS.SELF_UNASSIGN_CMD),
    assign_user_cmd: getInput(INPUTS.ASSIGN_USER_CMD),
    unassign_user_cmd: getInput(INPUTS.UNASSIGN_USER_CMD),
  };
}
