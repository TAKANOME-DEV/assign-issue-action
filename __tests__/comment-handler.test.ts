/* eslint-disable filenames/match-regex */
import * as core from '@actions/core';
import * as github from '@actions/github';
import commentHandler from '../src/lib/comment-handler';
import helpers from '../src/lib/helpers';

describe('command-handler', () => {
  let client: any;

  beforeAll(() => {
    jest.spyOn(core, 'setFailed');

    client = github.getOctokit('_');

    Object.assign(process.env, helpers.getDefaultValues());
  });

  beforeEach(() => {
    process.env.INPUT_GITHUB_TOKEN = 'token';
    process.env.INPUT_REQUIRED_LABEL = 'Not-Assigned';
    github.context.payload.issue!.labels = [{ name: 'Not-Assigned' }];
    github.context.payload.issue!.assignee = '';
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('exits early if token is not provided', async () => {
    process.env.INPUT_GITHUB_TOKEN = '';
    await commentHandler(core, github);
    expect(core.setFailed).toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith(
      `Missing required input: token = `
    );
  });

  it('exits early if required_label is not provided', async () => {
    process.env.INPUT_REQUIRED_LABEL = '';
    await commentHandler(core, github);
    expect(core.setFailed).toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith(
      `Missing required input: required_label = `
    );
  });

  it('exits early if required_label is set but not present', async () => {
    process.env.INPUT_REQUIRED_LABEL = 'Not-Assigned';
    github.context.payload.issue!.labels = [];
    await commentHandler(core, github);
    expect(core.setFailed).toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith(
      'Required label: [Not-Assigned] label not found in issue #1.'
    );
  });

  it('exits early if the issue is already assigned', async () => {
    github.context.payload.issue!.assignee = 'Saladin';
    await commentHandler(core, github);
    expect(core.setFailed).toHaveBeenCalled();
    expect(core.setFailed).toHaveBeenCalledWith(
      'Issue #1 is already assigned to @Saladin'
    );
  });

  it('assigns the user if every input is set', async () => {
    await commentHandler(core, github);
    expect(core.setFailed).toHaveBeenCalledTimes(4);
    expect(client.rest.issues.addAssignees).toHaveBeenCalled();
    expect(client.rest.issues.addAssignees).toHaveBeenCalledWith({
      owner: 'TAKANOME-DEV',
      repo: 'testing',
      issue_number: 1,
      assignees: ['John'],
    });
  });

  it('add "Is-Assigned" label if user is assigned', async () => {
    await commentHandler(core, github);
    expect(core.setFailed).toHaveBeenCalledTimes(4);
    expect(client.rest.issues.addLabels).toHaveBeenCalled();
    expect(client.rest.issues.addLabels).toHaveBeenCalledWith({
      owner: 'TAKANOME-DEV',
      repo: 'testing',
      issue_number: 1,
      labels: ['Is-Assigned'],
    });
  });

  it('uses a custom assigned_comment message', async () => {
    await commentHandler(core, github);
    expect(core.setFailed).toHaveBeenCalledTimes(4);
    expect(client.rest.issues.createComment).toHaveBeenCalled();
    expect(github.context.payload.comment).toMatchSnapshot();
  });
});
