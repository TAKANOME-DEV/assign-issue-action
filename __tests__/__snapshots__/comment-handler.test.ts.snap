// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command-handler uses a custom assigned_comment message 1`] = `
Object {
  "body": "/assign-to-me",
  "html_url": "https://github.com/TAKANOME-DEV/testing",
  "user": Object {
    "login": "John",
  },
}
`;
